### PR TITLE
GDPR checkout improvements

### PR DESCRIPTION
**Pull request by mikejolley**
This PR contains a series of adjustments to help store owners make their checkout GDPR compliant, or reduce their exposure by collecting less personal data.

I'm submitting this for feedback/review. There is one item outstanding that we can do later when WordPress core is ready and thats including some default privacy policy text. WordPress is working on a privacy policy page solution which we can use when it's ready.

## Improved the appearance of "descriptions"

Checkout fields can have descriptions added which show up below the fields. This is not actually used by core, but some plugins such as Checkout Field Editor expose this. I envision description could be useful for just-in-time privacy notices on certain fields, so what I've done is improved the appearance of these by:

- Styling them to look like 'info' notices
- Making them appear when the field gains focus, so all notices are not shown at once
- Supporting HTML inside the description.

Here is what that looks like:

slack - a8c 2018-04-06 13-00-18

## Allowed for custom terms and conditions text, and control over the checkbox + label

I've added some additional options to the customiser to add some policy text to the checkout. This is independent of the checkbox and be shown regardless.

The terms checkbox is now optional and the label can also be controlled. To maintain backwards compatibility, this is on by default if the user already setup a terms page.

terms

## Allowed for non-critical checkout fields to be disabled, or hidden.

We collect *phone*, *company*, and *address 2* on the checkout but I don't think they are commonly required on most checkouts. Company for example is not really useful outside of b2b, and phone is not needed to fulfil orders. If users don't need this personal data, we can add a simple toggle to disable it (this is instead of the snippets we recommend now).

customize checkout local wordpress dev 2018-04-06 12-33-18

Quite simply, this allows the user to switch between required, optional, and hidden.

## Control whether required or optional fields are labelled

Because of the above change, it's likely most fields will be required and so it may make more sense to switch the labelling rather than mark all fields. To that end, I've added this option:

monosnap 2018-04-06 12-35-01

If you label required fields you'll see this:

slack - a8c 2018-04-06 12-35-27

If you label optional fields you'll see this:

slack - a8c 2018-04-06 12-35-39

### CODE REVIEW

**Comment by mikejolley** [comment]
When that is merged we can include default text which utilizes that. This area does accept regular html links also.

**Comment by willgorham** [comment]
@allendav touched on this, but I have had multiple stores that requested separate links to both Privacy and T+C pages in the terms agreement text. It would be great if that capability could be incorporated somehow.

File: templates/checkout/terms.php
Comment (claudiulodro) [inline templates/checkout/terms.php]: I believe this needs some indenting.

**Comment by claudiulodro** [comment]
Looks good! I like these new features.

Left a couple small questions, but the code looks solid and all backwards-compatibility looks handled correctly.

File: templates/checkout/terms.php
Comment (mikejolley) [reply templates/checkout/terms.php]: Yes indeed

**Comment by mikejolley** [comment]
@claudiulodro Took care of your feedback - made it populate the cart with a dummy product so the checkout shows up.

Also tweaked the options somewhat to simplify the panel - I've required a design review from @jameskoster. @warrendholmes we agreed "checkout" would be fine for the panel name.

Final thing; I've integrated the WP privacy page coming in the next update into the customiser and included some default wording to include if the page is set. Looks like this:

slack - a8c 2018-04-11 16-45-54

Getting this text legal approved before merge :)

**Comment by claudiulodro** [comment]
New changes look good. 👍 

One new bug (?) I noticed is that the Terms and Conditions textarea doesn't seem to do anything any more. No matter what I type into it I get the default message. I think it's just an option key mismatch (`woocommerce_checkout_terms_and_conditions_text` => `woocommerce_checkout_terms_and_conditions_checkbox_text `)

screen shot 2018-04-11 at 9 52 06 am

**Review by jameskoster — COMMENTED** [review]
Shared these thoughts in Slack already but putting them here for posterity.

* I’d flip these https://cloudup.com/clB-dSNN-aK and probably hide the text input if the checkbox is disabled (if possible)
* In that input, the placeholder includes a shortcode but you can’t see it without inspecting source which obviously nobody will do. Instead of a placeholder can’t we just add this as a default value. Otherwise consider making it a textarea.
* The required fields asterisk option isn't working currently. Storefront theme active.

**Comment by claudiulodro** [comment]
Code looks good and I like the changes made. 

One question, is it on purpose that now you have both `*` and `(optional)` on the checkout fields when required stars are enabled?

screen shot 2018-04-16 at 10 45 12 am

**Comment by mikejolley** [comment]
@claudiulodro yes. I did that on purpose since it's only a few fields by default.

**Review by claudiulodro — APPROVED** [review]

**Comment by nicholasbown** [comment]
Hi! So in this update will there be an option for GDPR opt-in where there are two checkboxes, yes or no, where one has to be checked forcing the user to make a decision? At the moment it's just one. I've seen a case study where two checkboxes is +40% conversion to opting, hence my question.

**Comment by mikejolley** [comment]
There is no checkbox required for GDPR

**Comment by globalee** [comment]
@mikejolley Hi Mike , is it possible to have optional check boxes at checkout where customer can exspilcitly provide concent.

Example.  (free text to set a tick box for examlple: over 18, they consent they are over 18 or the gaurdian of someone if they are purchasing on behalf of someone who is under this age.  whereby this is written to db. I sell helth products and some require specific consent.

I have some code that im using atm but it dosent write to DB, but does stop customer from checking out  from buying specific products UNLESS they tick the box. 

This is the bones of it its just missing the write to DB.  is this possible to have an option for a FREE tick box(s) specified. 

`/**
 * Check if a specific product category is in the cart
 */
function wc_ninja_category_is_in_the_cart() {
	// Add your special category slugs here
	$categories = array( 'metabolic-packs', 'minerals', 'glandulars', 'vitamins', 'combination formulas', 'digestive aids' );

	// Products currently in the cart
	$cart_ids = array();

	// Categories currently in the cart
	$cart_categories = array();

	// Find each product in the cart and add it to the $cart_ids array
	foreach( WC()->cart->get_cart() as $cart_item_key => $values ) {
		$cart_product = $values['data'];
		$cart_ids[]   = $cart_product->id;
	}

	// Connect the products in the cart w/ their categories
	foreach( $cart_ids as $id ) {
		$products_categories = get_the_terms( $id, 'product_cat' );

		// Loop through each product category and add it to our $cart_categories array
		foreach ( $products_categories as $products_category ) {
			$cart_categories[] = $products_category->slug;
		}
	}

		// If one of the special categories are in the cart, return true.
	if ( ! empty( array_intersect( $categories, $cart_categories ) ) ) {
		return true;
	} else {
		return false;
	}
}
//add_filter( 'woocommerce_review_order_before_submit' , 'wc_ninja_remove_checkout_field', 9);
/**
 * Conditionally remove a checkout field based on products in the cart
 */
function wc_ninja_remove_checkout_field( $fields ) {
	if ( ! wc_ninja_category_is_in_the_cart() ) {
		unset( $fields['bbloomer_add_checkout_tickbox'] );
	}

	return $fields;
}

add_action('woocommerce_review_order_before_submit', 'bbloomer_add_checkout_tickbox', 9);
  
function bbloomer_add_checkout_tickbox() {
  if( wc_ninja_category_is_in_the_cart() ){
?>
 
<p class="form-row terms">
<input type="checkbox" class="input-checkbox" name="deliverycheck" id="deliverycheck" />
<label for="deliverycheck" class="checkbox">By purchasing products on this site, I consent that I am over 18 or the guardian of someone who is below this age.<span class="required">*</span></label>
</p>
 
<?php
}
}


// Show notice if customer does not tick
  
add_action('woocommerce_checkout_process', 'bbloomer_not_approved_delivery');

function bbloomer_not_approved_delivery() {
    if ( wc_ninja_category_is_in_the_cart() && ! (int) isset( $_POST['deliverycheck'] ) ) {
        wc_add_notice( __( '<strong>Please Acknowledge</strong> the terms of sale' ), 'error' );
    }
}`

**Comment by mikejolley** [comment]
We're not going to build in a solution for niche use cases like that. It's possible with custom code and our hooks already.

**Pull request merged**
